### PR TITLE
Explaining regression loss function for data frame analytics jobs

### LINKED ISSUE #330: ML explanation for different regression loss functions
I want to add a Jupyter notebook explaining different regression loss function implemented in Elasticsearch 7.8.  It will give more details and insights than it is possible in the scope of the Concepts documentation.

### PR DESCRIPTION
**Pull request by valeriy42**
This PR adds a notebook explains the properties of different loss functions available for the machine learning data frame analytics regression jobs and give pointers on when to choose which loss function.

It intends to be linked in the "Concepts" section of the documentation and provide users with more insights than it is possible in the scope of "Concepts".

Closes #330.

### CODE REVIEW

File: Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+    "they would work best in different scenarios. Let's look at what loss function works best for which case. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Mean squared error (MSE)\n",
+    "\n",
+    "[MSE](https://en.wikipedia.org/wiki/Mean_squared_error) if the most commonly used loss function. It works sufficiently well in many scenarios and you should try it if you are unsure or you don't know much about your data. \n",
```
Comment (tveasey) [inline Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb:33]: typo: 
```suggestion
    "[MSE](https://en.wikipedia.org/wiki/Mean_squared_error) is the most commonly used loss function. It works sufficiently well in many scenarios and you should try it if you are unsure or you don't know much about your data. \n",
```

File: Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+    "mse_plot.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The problem with the squared error is that it makes small errors (< 1.0) even smaller and large large errors (>1.0) disproportionately larger.\n",
+    "\n",
+    "This means that if you have a few outliers in your data, i.e. due to a large measurement error, the model will put useless effort to fit these data points and other points will fall by the wayside. \n",
```
Comment (tveasey) [inline Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb:72]: I think I'd capture that this will often degrade accuracy for the majority inliers:
```suggestion
    "This means that if you have a few outliers in your data, i.e. due to a large measurement error, the model will put useless effort to fit these data points often degrading the accuracy of the majority of predictions as a result. \n",
```

File: Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+    "\n",
+    "So what you often want to do is instead of fitting the mean **squared** error to fit the mean **absolute** error (MAE)\n",
```
Comment (tveasey) [inline Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb:74]: I think I'd rephrase this to say that in this case, i.e. when you suspect outliers, since we're saying by default use MSE:
```suggestion
    "In these cases what you would usually want to do is instead of fitting the mean **squared** error to fit the mean **absolute** error (MAE)\n",
```

Comment (tveasey) [reply Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb:74]: It might also be worth saying you can check the model for outliers after fitting by plotting a histogram of prediction errors to see if it might be worth changing loss.

File: Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "\n",
+    "The distribution of the values has a \"bump\" for smaller values and then a long tail with a small number of very large values. \n",
+    "\n",
+    "Naturally, when we want to predict the target values, we want to allow for an error depending on the magnitude of the target. It’s ok to miss a population of a metropolitan area by 10 000, but for a small town it is unforgivable. In this case MSLE can be more suitable for the regression job than MSE. As the name suggests, MSLE minimizes the quadratic difference between the logarithms of the actual value and the prediction\n",
+    "$$L(a_i, p_i) = (\\log(a_i + t) - \\log(p_i + t))^2$$\n",
+    "\n",
+    "Qualitatively speaking, MSLE tends to lead to lower error for small targets and for higher error on large targets than MSE. This trade-off can be controlled by adjusting an offset parameter $t$ which is set to 1.0 by default. By increasing this parameter we influence the “transition point” at which we go from minimizing quadratic error to minimizing quadratic log error. \n",
```
Comment (tveasey) [inline Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb:254]: This is all good. I think it is worth saying that in the log regime it is effectively targeting relative error, i.e.:
`log((1 + 1100) / (1 + 1000))^2` is about the same as `log((1 + 11000) / (1 + 10000))^2` so missing a value by 10% is penalised equivalently whether the actual error is 100 or 1000. Contrasting with MSE, the latter case costs 100x more.

File: Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+   "metadata": {},
+   "source": [
+    "Setting the offset parameter to 1.0 or 100.0 undershoots and overshoots this \"transition point\". Both lead to similar results with greater residuals for small targets and lower residuals for large targets compared to the offset parameter of 10.0. However, trying to \"nail\" the correct transition point super exactly is not worth it: setting offset parameter to 6.0 or 16.0  won't change the results significantly compared to 10.0.\n",
+    "\n",
+    "### Conclusion\n",
+    "\n",
+    "To use MSLE with a certain parameter offset $t$ in your data frame analytics regression job, you can specify parameter `loss_function` (set it to `msle`) and `loss_function_parameter`. Try it, if you have one of the [numerous examples](https://en.wikipedia.org/wiki/Log-normal_distribution#Occurrence_and_applications) of positive target values which scale over a large scale. You can leave the `loss_function_parameter` as it is, but if you want to experiment with different offset values, we recommend to set the value somewhere at the 80th percentile of the target distribution and then increase by an order of magnitude and look at the results qualitatively."
```
Comment (tveasey) [inline Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb:390]: This is good! It might also be worth saying that there should be strong evidence for positive skew: Huber would do better if there are symmetric heavy tails. For example, either measure the skewness or plot the histogram and check for the heavy tail. This can also be done after fitting the model with MSE: for example, checking non-normality of the prediction errors.

File: Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+    "qvalues = np.quantile(yexp, np.linspace(0, 1.0, 11))\n",
+    "for q, val in zip(quantiles, qvalues):\n",
+    "    print(\"Percentile {:.0f}\\tvalue {:.2f}\".format(q*100, val))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Setting the offset parameter to 1.0 or 100.0 undershoots and overshoots this \"transition point\". Both lead to similar results with greater residuals for small targets and lower residuals for large targets compared to the offset parameter of 10.0. However, trying to \"nail\" the correct transition point super exactly is not worth it: setting offset parameter to 6.0 or 16.0  won't change the results significantly compared to 10.0.\n",
```
Comment (tveasey) [inline Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb:386]: 👍

File: Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb (hunk 1 above)
Comment (valeriy42) [reply Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb:33]: Fixed

File: Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb (hunk 2 above)
Comment (valeriy42) [reply Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb:72]: fixed

File: Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb (hunk 3 above)
Comment (valeriy42) [reply Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb:74]: I also added a section on residual distrbution

File: Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb (hunk 5 above)
Comment (valeriy42) [reply Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb:390]: Updated

File: Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb (hunk 4 above)
Comment (valeriy42) [reply Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb:254]: Added

**Comment by valeriy42** [comment]
Thank you @tveasey for your comments. I tried to integrate all your suggestions. Can you please take another look when you have time?

File: Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb [hunk 7]
```diff
@@ -248,17 +316,19 @@
...
+    "Naturally, when you want to predict the target values, you want to allow for an error depending on the magnitude of the target. It’s ok to miss a population of a metropolitan area by 10 000, but for a small town it is unforgivable. In this case MSLE can be more suitable for the regression job than MSE. As the name suggests, MSLE minimizes the quadratic difference between the logarithms of the actual value and the prediction\n",
     "$$L(a_i, p_i) = (\\log(a_i + t) - \\log(p_i + t))^2$$\n",
     "\n",
-    "Qualitatively speaking, MSLE tends to lead to lower error for small targets and for higher error on large targets than MSE. This trade-off can be controlled by adjusting an offset parameter $t$ which is set to 1.0 by default. By increasing this parameter we influence the “transition point” at which we go from minimizing quadratic error to minimizing quadratic log error. \n",
+    "For example, $\\log\\left((1 + 1100) / (1 + 1000)\\right)^2$ is about the same as $\\log((1 + 11000) / (1 + 10000))^2$ so missing a value by 10\\% is penalised equivalently whether the actual error is 100 or 1000. At the same time, with MSE the latter case costs 100x more.\n",
```
Comment (tveasey) [inline Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb:322]: Maybe to carry on the earlier example you could use `log((1 + 110000) / (1 + 100000)` is about the same as `log((1 + 11000) / (1 + 10000))^2`, i.e. this relates directly back to the population example.
